mlflow/tracing/processor/mlflow.py

```diff
@@ -86,10 +86,26 @@ def on_start(self, span: OTelSpan, parent_context: Optional[Context] = None):
         span.set_attribute(SpanAttributeKey.REQUEST_ID, json.dumps(request_id))
 
     def _start_trace(self, span: OTelSpan, start_time_ns: Optional[int]) -> TraceInfo:
+        from mlflow.tracking.fluent import _active_run_stack
+
         experiment_id = get_otel_attribute(span, SpanAttributeKey.EXPERIMENT_ID)
         metadata = {TRACE_SCHEMA_VERSION_KEY: str(TRACE_SCHEMA_VERSION)}
         # If the span is started within an active MLflow run, we should record it as a trace tag
-        if run := mlflow.active_run():
+        # Note `MLflow.active_run()` can only get thread-local active run,
+        # but tracing routine might be applied to model inference worker threads
+        # (e.g. langchain model batch inference),
+        # and we only support global mode tracing,
+        # so the following code checks all active runs in all threads to get the latest active run
+        # as the tracing source run.
+        latest_active_run = None
+        for active_run_stack in _active_run_stack._value_dict.values():
```

Comment:
Is there any chance `_active_run_stack` will also be updated to thread local? It seems the stack is used for determining parent run in nested case, then it might be a scope of thread-safety?
https://github.com/WeichenXu123/mlflow/blob/5da87886ab760e61b5add6e175b4490397378c80/mlflow/tracking/fluent.py#L396-L398

Reply:
before this commit https://github.com/mlflow/mlflow/commit/487db1f6c9b62c40ccfcba76c25defdabbcc4acc

all threads shares the same `_active_run_stack`, i.e., even it is a stack, all threads can push runs into it, it is buggy in multiple thread case.